src/main/java/com/faforever/client/map/generator/MapGeneratorService.java

```diff
@@ -107,43 +115,87 @@ private void deleteGeneratedMaps() {
+    ByteBuffer seedBuffer = ByteBuffer.allocate(8);
+    seedBuffer.putLong(seedGenerator.nextLong());
+    String seedString = Base64.getEncoder().encodeToString(seedBuffer.array());
+    return generateMap(generatorVersion, seedString);
+  }
+
+  public CompletableFuture<String> generateMap(byte[] optionArray) {
+    return generateMap(generatorVersion, optionArray);
+  }
+
+  public CompletableFuture<String> generateMap(String version, byte[] optionArray) {
+    return generateMap(new ComparableVersion(version), optionArray);
+  }
+
+  public CompletableFuture<String> generateMap(ComparableVersion version, byte[] optionArray) {
+    ByteBuffer seedBuffer = ByteBuffer.allocate(8);
+    seedBuffer.putLong(seedGenerator.nextLong());
+    String seedString = Base64.getEncoder().encodeToString(seedBuffer.array());
+    String optionString = Base64.getEncoder().encodeToString(optionArray);
+    return generateMap(version, seedString + '_' + optionString);
   }
 
   @VisibleForTesting
-  protected String queryNewestVersion() {
+  @Cacheable(CacheNames.MAP_GENERATOR)
+  public ComparableVersion queryMaxSupportedVersion() {
+    ComparableVersion version = new ComparableVersion("");
+    ComparableVersion minVersion = new ComparableVersion(String.valueOf(clientProperties.getMapGenerator().getMinSupportedMajorVersion()));
+    ComparableVersion maxVersion = new ComparableVersion(String.valueOf(clientProperties.getMapGenerator().getMaxSupportedMajorVersion() + 1));
+
     RestTemplate restTemplate = new RestTemplate();
 
     LinkedMultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
     headers.add("Accept", "application/vnd.github.v3+json");
     HttpEntity<String> entity = new HttpEntity<>(null, headers);
 
-    ResponseEntity<String> response = restTemplate.exchange(clientProperties.getMapGenerator().getQueryLatestVersionUrl(), HttpMethod.GET, entity, String.class);
-    JsonElement jsonElement = new JsonParser().parse(response.getBody());
-    JsonObject mainObject = jsonElement.getAsJsonObject();
-
-    return mainObject.get("tag_name").getAsString();
+    ResponseEntity<List<GithubGeneratorRelease>> response = restTemplate.exchange(clientProperties.getMapGenerator().getQueryVersionsUrl(), HttpMethod.GET, entity, new ParameterizedTypeReference<List<GithubGeneratorRelease>>(){});
+    List<GithubGeneratorRelease> releases = response.getBody();
+    for (GithubGeneratorRelease release : releases) {
+      version.parseVersion(release.getTag_name());
+      if (version.compareTo(maxVersion) < 0 && minVersion.compareTo(version) < 0) {
+        return version;
+      }
+    }
+    throw new RuntimeException("No Valid Generator Version Found");
   }
 
   public CompletableFuture<String> generateMap(String mapName) {
     Matcher matcher = GENERATED_MAP_PATTERN.matcher(mapName);
     if (!matcher.find()) {
-      throw new IllegalArgumentException("Map name is not a generated map");
+      return CompletableFuture.supplyAsync(() -> {
```

Comment:
I think there is a static Method on the CompletableFuture class that does exactly that

Reply:
Used